File: modules/session/session-impl/src/main/java/com/sportconnect/session/entity/Session.java
package com.sportconnect.session.entity;

import com.sportconnect.session.api.dto.FeeType;
import com.sportconnect.session.api.dto.SessionStatus;
import com.sportconnect.session.api.dto.SessionType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.EnumType;
import jakarta.persistence.Enumerated;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.Table;
import jakarta.persistence.UniqueConstraint;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;
import java.util.UUID;

@Entity
@Table(name = "sessions", uniqueConstraints = {
        @UniqueConstraint(name = "unique_group_session_start", columnNames = {"group_id", "scheduled_start"})
})
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class Session {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "group_id")
    private Long groupId;

    @Enumerated(EnumType.STRING)
    @Column(name = "session_type", nullable = false, length = 30)
    private SessionType sessionType;

    @Column(name = "created_by", nullable = false)
    private UUID createdBy;

    @Column(name = "sport_id", nullable = false)
    private Long sportId;

    @Column(length = 200)
    private String title;

    @Column(columnDefinition = "TEXT")
    private String description;

    @Column(name = "location_id", nullable = false)
    private Long locationId;

    /** Free-text detail within the shared Location, e.g. "Court 3" or "North field" —
     * scoped to this one session, never written back to the shared Location row. */
    @Column(name = "location_note", length = 500)
    private String locationNote;

    @Column(name = "scheduled_start", nullable = false)
    private LocalDateTime scheduledStart;

    @Column(name = "scheduled_end_at")
    private LocalDateTime scheduledEndAt;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false, length = 20)
    @Builder.Default
    private SessionStatus status = SessionStatus.SCHEDULED;

    /** "Open slots" — informational/display only, never enforced by joinSession. 9999 is the
     * backfill sentinel for sessions predating this column (and auto-generated recurring
     * sessions, which have no capacity input) — see V040 migration. */
    @Column(nullable = false)
    @Builder.Default
    private Integer capacity = 9999;

    @Enumerated(EnumType.STRING)
    @Column(name = "fee_type", nullable = false, length = 10)
    @Builder.Default
    private FeeType feeType = FeeType.FREE;

    /** Meaningful only when feeType == FIXED; SessionServiceImpl clears this to null otherwise. */
    @Column(name = "fee_amount_vnd")
    private Long feeAmountVnd;

    /** Participants the creator already accounts for outside the app (e.g. a team they brought
     * with them) — added on top of the real JOINED SessionParticipant count when SessionServiceImpl
     * reports participantCount. Never negative; 0 for every row predating this column. */
    @Column(name = "initial_slot", nullable = false)
    @Builder.Default
    private Integer initialSlot = 0;

    /** false = joinSession puts non-invited joiners into PENDING, awaiting creator/owner-admin
     * approval. Sessions predating this column backfilled to true (see V041 migration) so their
     * existing instant-join behavior didn't silently change. */
    @Column(name = "auto_approve", nullable = false)
    @Builder.Default
    private Boolean autoApprove = false;

    @Column(name = "cancel_reason", length = 500)
    private String cancelReason;

    @Column(name = "cancelled_by")
    private UUID cancelledBy;

    @Column(name = "cancelled_at")
    private LocalDateTime cancelledAt;

    @CreationTimestamp
    @Column(name = "created_at", nullable = false, updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (!(o instanceof Session)) return false;
        Session session = (Session) o;
        return id != null && id.equals(session.getId());
    }

    @Override
    public int hashCode() {
        return getClass().hashCode();
    }
}

File: client/src/shared/types/session.ts
import type { Location } from '@/shared/types/location';

// Typed 1:1 against the real backend DTOs (modules/session/session-api/.../dto/) — SessionResponse,
// SessionParticipantResponse, SessionType, SessionStatus, ParticipantStatus. Lives in shared/ (not
// features/session/) because both the Matches page and the cross-page UpcomingMatches rail card
// need it — the same reasoning shared/types/rail.ts already documents for SportKey/SportProfile.

// TOURNAMENT/TRAINING are reserved backend enum values with no supporting logic yet
// (modules/session/session-impl/CLAUDE.md) — included here for type-completeness only.
export type SessionType = 'GROUP_RECURRING' | 'STANDALONE' | 'TOURNAMENT' | 'TRAINING';

export type SessionStatus = 'SCHEDULED' | 'ONGOING' | 'COMPLETED' | 'CANCELLED';

// SESSION-6: REQUESTED = self-initiated join awaiting creator/owner-admin approval (only reachable
// when the session's autoApprove is false). INVITED = pre-created at session creation from
// CreateSessionRequest.inviteeIds, resolved only by the invitee's own joinSession call (which
// bypasses the approval gate regardless of autoApprove) — no separate accept endpoint exists.
export type ParticipantStatus = 'JOINED' | 'LEFT' | 'REQUESTED' | 'INVITED';

// SESSION-5: capacity is display-only (never enforced by joinSession); feeAmountVnd is
// meaningful only when feeType is FIXED (null otherwise).
export type FeeType = 'FREE' | 'SPLIT' | 'FIXED';

export interface Session {
  id: number;
  groupId: number | null;
  sessionType: SessionType;
  createdBy: string;
  createdByFullName: string;
  sportId: number;
  sportName: string;
  title: string | null;
  description: string | null;
  location: Location;
  locationNote: string | null;
  scheduledStart: string; // ISO, LocalDateTime (no timezone)
  scheduledEndAt: string | null;
  status: SessionStatus;
  cancelReason: string | null;
  cancelledBy: string | null;
  cancelledByFullName: string | null;
  cancelledAt: string | null;
  /** Real JOINED SessionParticipant rows + initialSlot, computed backend-side — not a raw
   * participant-table count. */
  participantCount: number;
  capacity: number;
  feeType: FeeType;
  feeAmountVnd: number | null;
  /** Participants the creator already accounts for outside the app, folded into
   * `participantCount` above (backend-side, `SessionServiceImpl.mapToResponses`). */
  initialSlot: number;
  /** SESSION-6: false = non-invited joiners land in REQUESTED, awaiting creator/owner-admin
   * approval. An INVITED row always bypasses this gate regardless of this value. */
  autoApprove: boolean;
  /** Like state of this session's own SESSION_POST anchor (backend-side, SESSION-10's
   * "post-ship addition" — `POST/DELETE /sessions/{id}/like`). */
  likeCount: number;
  isLikedByCurrentUser: boolean;
  /** SESSION-9: the caller's own SessionParticipant row for this session, or null if they have
   * none. userFullName/userAvatarUrl are unset on it (the caller already knows their own
   * identity) — only id/sessionId/userId/status/rejectReason/createdAt are meaningful. Drives
   * the Join/Accept/Decline/Cancel/Leave action on both the session card and SessionDetailModal:
   * null or LEFT -> Join, INVITED -> Accept/Decline, REQUESTED -> Cancel, JOINED -> Leave. */
  callerParticipation: SessionParticipant | null;
  createdAt: string;
  updatedAt: string;
}

export interface SessionParticipant {
  id: number;
  sessionId: number;
  userId: string;
  userFullName: string;
  userAvatarUrl: string | null;
  status: ParticipantStatus;
  /** Set only by rejectParticipant — null otherwise, including for a LEFT row from a self-leave. */
  rejectReason: string | null;
  createdAt: string;
}
